Fix calculation of freshness lifetime in HttpResponse

arturobernalg opened the pull request:
This pull request adds a fix for the calculateFreshnessLifetime method in the HttpResponseCache class. The method calculates the freshness lifetime of a response based on the headers in the response and follows the algorithm for calculating the freshness lifetime described in RFC 7234, section 4.2.1.

The method takes into account the s-maxage, max-age, Expires, and Date headers as follows:

- If the s-maxage directive is present in the Cache-Control header of the response, its value is used as the freshness lifetime for shared caches, which typically serve multiple users or clients.

- If the max-age directive is present in the Cache-Control header of the response, its value is used as the freshness lifetime for private caches, which serve a single user or client.

- If the Expires header is present in the response, its value is used as the expiration time of the response. The freshness lifetime is calculated as the difference between the expiration time and the time specified in the Date header of the response.

- If none of the above headers are present or if the calculated freshness lifetime is invalid, a default value of 5 minutes is returned.

The existing implementation had a bug where it did not handle the case where the max-age directive was present in the Cache-Control header but the s-maxage directive was not. In this case, the method would incorrectly return the default freshness lifetime of 5 minutes instead of using the max-age value as the freshness lifetime. This pull request fixes the bug by adding a check for the s-maxage directive and returning the max-age value if the s-maxage directive is not present.

Motivation

The bug in the calculateFreshnessLifetime method could cause incorrect caching behavior, which could lead to degraded performance and unexpected results for clients. By fixing this bug, we ensure that the method correctly calculates the freshness lifetime of responses and that clients can rely on the cache to serve responses efficiently and accurately.

Changes Made

To fix the bug, we added a check for the s-maxage directive in the Cache-Control header and updated the return value of the method to use the max-age value if the s-maxage directive is not present.

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/ResponseCachingPolicy.java:
```diff
@@ -50,6 +51,29 @@
+     * <p>By default, this value is set to 300 seconds (5 minutes). Applications can customize this
+     * value as needed.</p>
+     */
+    private static final long DEFAULT_FRESHNESS_DURATION = 300L;
```
@arturobernalg This is actually the case where the use of `java.time.Duration` instead of plain `long` might be beneficial.

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/ResponseCachingPolicy.java:
```diff
@@ -320,4 +354,87 @@ private boolean from1_0Origin(final HttpResponse response) {
+     * implementing additional caching mechanisms as needed.
+     * </p>
+     *
+     * @param response the HTTP response for which to calculate the freshness lifetime
```
@arturobernalg We stopped referencing specific RFCs in javadocs. Those URLs tend to change and also need to be updated every time the code base gets revised against a newer set of RFCs.

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/ResponseCachingPolicy.java:
```diff
@@ -320,4 +354,87 @@ private boolean from1_0Origin(final HttpResponse response) {
+        long maxAge = -1;
+        long sharedMaxAge = -1;
+
+        for (final String headerValue : headerValues) {
```
@arturobernalg Please use `org.apache.hc.core5.util.Tokenizer` based parsers from HttpCore to parse header elements, see `AuthChallengeParser` for an example.
Please also consider what is going to happen if your code has to parse something like `max-age=boom`. Malformed header values must not cause a runtime exception.

ok2c replied:
@arturobernalg This is even worse then before. You should do all the parsing with `Tokenizer` and there should be no `split`, `trim` or `substring` and there should be almost no intermediate garbage. Please see how `Tokenizer`is being used to parse other complex headers, like `Set-Cookie` for instance. Please take a look at `RFC6265CookieSpec#parse`.

arturobernalg commented:
HI @ok2c 
All remarks changed and Refactor calculateFreshnessLifetime() to use Tokenizer for parsing Cache-Control header values.

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/ResponseCachingPolicy.java:
```diff
@@ -320,4 +360,100 @@ private boolean from1_0Origin(final HttpResponse response) {
+        // Check if Expires is present and use its value minus the value of the Date header
+        Instant expiresInstant = null;
+        Instant dateInstant = null;
+        if (response.containsHeader(HttpHeaders.EXPIRES)) {
```
@arturobernalg Please do not do that. Please do not use two method calls where a single method call and a null check should perfectly suffice.

arturobernalg commented:
@ok2c  WDYT?

arturobernalg commented:
> @arturobernalg That is much better, but you can still improve it further. Please note that your method executes the same logic twice inside the loop. Please optimize it. You may also want to create a separate test case just for the parser. Once this is done I will suggest further improvements.

HI @ok2c 
what about now.
TY

arturobernalg commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheControlHeaderParser.java:
```diff
@@ -0,0 +1,211 @@
+            }
+            final int valueDelim = buffer.charAt(cursor.getPos());
+            cursor.updatePos(cursor.getPos() + 1);
+            if (valueDelim != EQUAL_CHAR) {
```
@ok2c 
This is the only part that I'm not sure about. I don't think these conditions will ever happen. Otherwise, they would be controlled by the previous one.

arturobernalg replied:
its already cover here,
we really not need that validations
`   @Test
    public void testParseCookieEmptyValue() {
        final Header header = new BasicHeader("Cache-Control", "max-age=;");
        assertThrows(MalformedCookieException.class, () -> parser.parse(header));
    }`

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheControlHeaderParser.java:
```diff
@@ -0,0 +1,211 @@
+        } else {
+            final String s = header.getValue();
+            if (s == null) {
+                if (LOG.isWarnEnabled()) {
```
@arturobernalg No need for a warning here.

arturobernalg replied:
remove it

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheControlHeaderParser.java:
```diff
@@ -0,0 +1,211 @@
+
+        while (!cursor.atEnd()) {
+            final String name = tokenParser.parseToken(buffer, cursor, TOKEN_DELIMS);
+            if (name.isEmpty()) {
```
@arturobernalg Make sure your parser will not choke on `Cache-Control: , = blah,` value. Please add this test case.

arturobernalg replied:
handle by while (!cursor.atEnd())

ok2c replied:
@arturobernalg Are you sure. Anyway I cannot see a test case for it.

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheControlHeaderParser.java:
```diff
@@ -0,0 +1,211 @@
+                continue;
+            }
+            if (cursor.atEnd()) {
+                if (LOG.isWarnEnabled()) {
```
@arturobernalg Is this really so? What about `Cache-Control: no-cache` and similar? There can be several directives that do not have a parameter. Please remove or rewrite this warning.

arturobernalg replied:
remove it

arturobernalg commented:
all remarks solved

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheControlHeaderParser.java:
```diff
@@ -0,0 +1,205 @@
+            final int valueDelim = buffer.charAt(cursor.getPos());
+            cursor.updatePos(cursor.getPos() + 1);
+            if (valueDelim != EQUAL_CHAR) {
+                if (LOG.isWarnEnabled()) {
```
@arturobernalg Do we really need this warning? Is the header really invalid or is it a directive without a parameter?

ok2c commented on httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheControlHeaderParser.java:
```diff
@@ -0,0 +1,205 @@
+     * @return true if the header field is not recognized
+     * @throws MalformedCookieException if the value is malformed and cannot be parsed as a long integer
+     */
+    private boolean setHeader(final CacheControlHeader cacheControlHeader, final String name, final String value) throws MalformedCookieException {
```
@arturobernalg `MalformedCookieException` is clearly wrong here. This header has nothing to do with cookies. I also think the right thing to do here is to ignore values that cannot be interpreted as valid instead of throwing an exception. A log message at the DEBUG priority might be all right.

Another thing. Please  rename `CacheControlHeader` to `CacheControl` and make it nice and *immutable*. For that you may need to inline this method and create `CacheControl` instance at the last moment when exiting the `#parse` method.

arturobernalg commented:
@ok2c done?  the only thing that I didn't now what you mean was -->For that you may need to inline this method and create CacheControl instance at the last moment when exiting the #parse method.
but in anywise, seems everything ok.

ok2c commented:
> @ok2c done? 

@arturobernalg  No, not quite. You added `@Immutable` to `CacheControl` but it is not immutable. Its instance variables are not final. You need to make sure the instances of `CacheControl` cannot be mutated after construction.

arturobernalg commented:
> > @ok2c done?
> 
> @arturobernalg No, not quite. You added `@Immutable` to `CacheControl` but it is not immutable. Its instance variables are not final. You need to make sure the instances of `CacheControl` cannot be mutated after construction.

HI @ok2c

You are correct, adding @Immutable to CacheControl does not make it immutable by itself. I have made the necessary changes to make CacheControl immutable.

I declared the instance variables maxAge and sharedMaxAge as final, removed their setter methods, and modified the constructor to accept the values for these variables and initialize them with these values.

With these changes, instances of CacheControl cannot be mutated after construction, making it truly immutable. Please let me know if you have any further concerns or feedback.

ok2c reviewed (approved):
@arturobernalg Looks good now. I will commit your changes with some minor tweaks later today.

ok2c reviewed (changes requested):
@arturobernalg I asked you to look at this test case, did not I? What do you think is going to happen?
```
    @Test
    public void testGarbage() {
        final Header header = new BasicHeader("Cache-Control", ",,= blah,");
        final CacheControl cacheControl = parser.parse(header);
        assertEquals(0L, cacheControl.getMaxAge());
    }
```

ok2c commented:
@arturobernalg Please also remove `PARAM_DELIMITER` from the `CacheControlHeaderParser`. It is not applicable to the cache control directives. They do not have parameters.

arturobernalg commented:
@ok2c 

Tackled infinite loop caused by garbage input in testGarbage() test case
Removed unnecessary PARAM_DELIMITER from CacheControlHeaderParser

ok2c commented:
@arturobernalg There is a simpler, nicer fix. Wanna try to find it?

Bonus question: what should the value of max-age be with the header like `Cache-Control: no-cache, max-age=123`?

arturobernalg commented:
> @arturobernalg There is a simpler, nicer fix. Wanna try to find it?
> 
> Bonus question: what should the value of max-age be with the header like `Cache-Control: no-cache, max-age=123`?

I just rt

> @arturobernalg There is a simpler, nicer fix. Wanna try to find it?
> 
> Bonus question: what should the value of max-age be with the header like `Cache-Control: no-cache, max-age=123`?
For what I understand a cache MUST NOT use a response to satisfy a subsequent request without successful revalidation with the origin server if the Cache-Control header field includes the "no-cache" directive. However, it MAY use the response to satisfy a subsequent request if the Cache-Control header field contains a "max-age" directive with a non-zero value.

In the given header "Cache-Control: no-cache, max-age=123", the "no-cache" directive takes precedence over the "max-age" directive, but the presence of "max-age" means that the response can be used to satisfy a subsequent request after 123 seconds have elapsed, assuming successful revalidation with the origin server. So the value of max-age should be 123 seconds.

Regarding the simple fix, we can return the CacheControl object since we have already extracted the relevant caching directives from the Cache-Control header by this point.

ok2c commented:
@arturobernalg I am going to merge the PR as is and make some small changes to it. The current handling of with different (conflicting) cache control directives should be fixed, however. In the simplest form the `max-age` value should be ignored if the `no-cache` directive is present.

ok2c commented:
@arturobernalg I now understood your response better. It turns out that now the same `Cache-Control` header gets parsed twice, once by `#isExplicitlyNonCacheable` and again by `#calculateFreshnessLifetime`, which is bad and must be fixed. The parsing code should be extracted from `#calculateFreshnessLifetime` and enhanced to include the main cache control directive that the `#isExplicitlyNonCacheable` could use to make its decision:

```
        if (isExplicitlyNonCacheable(response)) {
            LOG.debug("Response is explicitly non-cacheable");
            return false;
        }

        // calculate freshness lifetime
        final Duration freshnessLifetime = calculateFreshnessLifetime(response);
        if (freshnessLifetime.isNegative() || freshnessLifetime.isZero()) {
            if (LOG.isDebugEnabled()) {
                LOG.debug("Freshness lifetime is invalid");
            }
            return false;
        }
```

arturobernalg commented:
> LOG.debug

@ok2c  I'll take a look
